src/service/cid-impl.js

```diff
@@ -230,26 +230,33 @@ export class Cid {
       }
       return getOrCreateCookie(this, getCidStruct, persistenceConsent);
     }
-    if (this.cacheCidApi_.isSupported()) {
-      const apiKey = this.isScopeOptedIn_(scope);
-      if (!apiKey) {
-        return /** @type {!Promise<?string>} */ (Promise.resolve(null));
-      }
-      return this.cacheCidApi_.getScopedCid(scope);
-    }
+
     return this.viewerCidApi_.isSupported().then(supported => {
       if (supported) {
         const apiKey = this.isScopeOptedIn_(scope);
         return this.viewerCidApi_.getScopedCid(apiKey, scope);
       }
-      return getBaseCid(this, persistenceConsent)
-          .then(baseCid => {
-            return Services.cryptoFor(this.ampdoc.win).sha384Base64(
-                baseCid + getProxySourceOrigin(url) + scope);
-          });
+
+      if (this.cacheCidApi_.isSupported() && this.isScopeOptedIn_(scope)) {
+        return this.cacheCidApi_.getScopedCid(scope).then(scopedCid => {
+          if (scopedCid) {
+            return scopedCid;
+          }
+          return this.scopeBaseCid_(persistenceConsent, scope, url);
+        });
+      }
```

Comment:
Why move `cacheCidApi_` logic here? Native aGSA doesn't have a viewer, right?

Reply:
yep, so `supported` is `false` in this case.

Moving here is mainly to simplify the async code a bit. It's safe because `this.viewerCidApi_.isSupported()` and `this.cacheCidApi_.isSupported()` are mutually exclusive.

Reply:
Tried, but seems complicated in this async code style. Any good techniques here ?

Reply:
Can we just revert this nesting (move this cache CID block back above the viewer CID block)?

As long as cache CID and viewer CID can't both be supported, that should be fine. Or maybe I'm missing something.

Reply:
you will introduce another nesting:
```
if (this.cacheCidApi_.isSupported()) {
  if (this.isScopeOptedIn_(scope)) {
      return...
  } 
  return ...
}
```

and you will also repeat the `this.scopeBaseCid_` a 3rd time.